linera-core/src/client/mod.rs

```diff
@@ -3492,34 +3492,40 @@ impl<Env: Environment> ChainClient<Env> {
         maybe_info.map_or(BlockHeight::ZERO, |info| info.next_block_height)
     }
 
+    /// Returns the next height we expect to receive from the given sender chain, according to the
+    /// local inbox.
+    #[instrument(level = "trace")]
+    async fn local_next_height_to_receive(
+        &self,
+        origin: ChainId,
+    ) -> Result<BlockHeight, ChainClientError> {
+        let chain = self.chain_state_view().await?;
+        Ok(match chain.inboxes.try_load_entry(&origin).await? {
+            Some(inbox) => inbox.next_block_height_to_receive()?,
+            None => BlockHeight::ZERO,
+        })
+    }
+
     #[instrument(level = "trace", skip(remote_node, local_node, notification))]
     async fn process_notification(
         &self,
         remote_node: RemoteNode<Env::ValidatorNode>,
         mut local_node: LocalNodeClient<Env::Storage>,
         notification: Notification,
-    ) {
+    ) -> Result<(), ChainClientError> {
         match notification.reason {
             Reason::NewIncomingBundle { origin, height } => {
-                if self.local_next_block_height(origin, &mut local_node).await > height {
+                if self.local_next_height_to_receive(origin).await? > height {
                     debug!(
                         chain_id = %self.chain_id,
                         "Accepting redundant notification for new message"
                     );
-                    return;
+                    return Ok(());
                 }
-                if let Err(error) = self
-                    .find_received_certificates_from_validator(remote_node)
-                    .await
-                {
-                    error!(
-                        chain_id = %self.chain_id,
-                        "NewIncomingBundle: Fail to process notification: {error}"
-                    );
-                    return;
-                }
-                if self.local_next_block_height(origin, &mut local_node).await <= height {
-                    error!(
+                self.find_received_certificates_from_validator(remote_node)
+                    .await?;
+                if self.local_next_height_to_receive(origin).await? <= height {
```

Comment:
I think the `if` here is unnecessary. We just checked above if it's `> height` so the alternative it's `<=`.

Reply:
Ah, I didn't think that it might change it.